ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/RangeActionFilter.java

```diff
@@ -99,24 +100,63 @@ public void filterTsos() {
         Set<String> activatedTsos = leaf.getActivatedNetworkActions().stream().map(RemedialAction::getOperator).collect(Collectors.toSet());
         activatedTsos.addAll(appliedRangeActions.stream().map(RemedialAction::getOperator).collect(Collectors.toSet()));
 
-        Set<String> tsosToKeep = new HashSet<>(activatedTsos);
+        Set<String> tsosToKeep = sortTsosToKeepBySensitivityAndGroupId(activatedTsos, maxTso);
 
-        List<RangeAction> rangeActionsSortedBySensitivity = rangeActionsToOptimize.stream()
-                .sorted((ra1, ra2) -> -compareAbsoluteSensitivities(ra1, ra2, leaf.getMostLimitingElements(1).get(0), leaf))
-                .collect(Collectors.toList());
-        for (RangeAction rangeAction : rangeActionsSortedBySensitivity) {
-            if (tsosToKeep.size() >= maxTso) {
-                break;
-            }
-            tsosToKeep.add(rangeAction.getOperator());
-        }
         Set<RangeAction> rangeActionsToRemove = rangeActionsToOptimize.stream().filter(rangeAction -> !tsosToKeep.contains(rangeAction.getOperator())).collect(Collectors.toSet());
         if (!rangeActionsToRemove.isEmpty()) {
             LOGGER.info("{} range actions have been filtered out in order to respect the maximum allowed number of tsos", rangeActionsToRemove.size());
             rangeActionsToOptimize.removeAll(rangeActionsToRemove);
         }
     }
 
+    Set<String> sortTsosToKeepBySensitivityAndGroupId(Set<String> activatedTsos, int maxTso) {
+        List<RangeAction> rangeActionsSortedBySensitivity = rangeActionsToOptimize.stream()
+                .sorted((ra1, ra2) -> -compareAbsoluteSensitivities(ra1, ra2, leaf.getMostLimitingElements(1).get(0), leaf))
+                .collect(Collectors.toList());
```

Comment:
you can create a function to encapsulate this + the same call in method removeRangeActionsBySensitivityAndGroupId (but with different sets of range actions)